File: packages/ruby-sdk/lib/opensend/resources/domains.rb
# frozen_string_literal: true

module OpenSend
  module Resources
    class Domains
      def initialize(http)
        @http = http
      end

      # POST /api/domains
      def create(params)
        @http.post("/api/domains", stringify(params))
      end

      # GET /api/domains
      def list
        @http.get("/api/domains")
      end

      # GET /api/domains/:id
      def get(id)
        @http.get("/api/domains/#{id}")
      end

      # PATCH /api/domains/:id
      def update(id, params)
        @http.patch("/api/domains/#{id}", stringify(params))
      end

      # POST /api/domains/:id/verify
      def verify(id)
        @http.post("/api/domains/#{id}/verify")
      end

      # DELETE /api/domains/:id
      def delete(id)
        @http.delete("/api/domains/#{id}")
      end

      private

      def stringify(params)
        return params unless params.is_a?(Hash)

        params.transform_keys(&:to_s)
      end
    end
  end
end

File: packages/go-sdk/domains.go
package opensend

import (
	"context"
	"net/http"
)

// DomainsClient provides access to the /api/domains endpoints.
type DomainsClient struct {
	c *Client
}

// Create adds a new domain via POST /api/domains.
func (d *DomainsClient) Create(ctx context.Context, req CreateDomainRequest) (*DomainResponse, error) {
	var out DomainResponse
	if err := d.c.doJSON(ctx, http.MethodPost, "/api/domains", req, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List retrieves all domains via GET /api/domains.
func (d *DomainsClient) List(ctx context.Context) (*DomainListResponse, error) {
	var out DomainListResponse
	if err := d.c.doJSON(ctx, http.MethodGet, "/api/domains", nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get retrieves a single domain by ID via GET /api/domains/{id}.
func (d *DomainsClient) Get(ctx context.Context, id string) (*DomainResponse, error) {
	var out DomainResponse
	if err := d.c.doJSON(ctx, http.MethodGet, "/api/domains/"+id, nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update partially updates a domain via PATCH /api/domains/{id}.
func (d *DomainsClient) Update(ctx context.Context, id string, req UpdateDomainRequest) (*DomainUpdateResponse, error) {
	var out DomainUpdateResponse
	if err := d.c.doJSON(ctx, http.MethodPatch, "/api/domains/"+id, req, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Verify triggers DNS verification for a domain via POST /api/domains/{id}/verify.
func (d *DomainsClient) Verify(ctx context.Context, id string) (*DomainResponse, error) {
	var out DomainResponse
	if err := d.c.doJSON(ctx, http.MethodPost, "/api/domains/"+id+"/verify", nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a domain via DELETE /api/domains/{id}.
func (d *DomainsClient) Delete(ctx context.Context, id string) (*DeleteDomainResponse, error) {
	var out DeleteDomainResponse
	if err := d.c.doJSON(ctx, http.MethodDelete, "/api/domains/"+id, nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
